keps/sig-storage/1855-csi-driver-service-account-token/README.md

```diff
@@ -0,0 +1,252 @@
+    because the CSI driver should have a `hostPath` volume for the `pods`
+    subdirectory to read the token.
+3.  The audience of the token is defaulted to kube apiserver.
+4.  The token is not guaranteed to be available (e.g.
+    `automountServiceAccountToken=false`).
+
+### User stories
+
+- HashiCorp Vault provider for secret store CSI driver requires service
+  account token of the pods they are mounting secrets at to authenticate to
+  Vaults. The provisioned secrets also have given TTL in Vault, so it is
+  necessary get tokens after the initial mount.
+- Cert manager CSI dirver will create CertificateRequests on behalf of the
+  pods.
+- Amazon EFS CSI driver wants the service account tokens of pods to exchange
+  for AWS credentials.
+
+## Proposal
+
+### Goals
+
+- Allow CSI driver to request audience-bounded service account tokens of pods
+  from kubelet to `NodePublishVolume`.
+- Provide an option to re-execute `NodePublishVolume` in a best-effort manner.
+
+### Non-Goals
+
+- Other CSI calls may not acquire pods' service account tokens via this
+  feature.
+- Failed re-execution of `NodePublishVolume` will not unmount volumes.
+
+### API Changes
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountToken *ServiceAccountToken
+}
+
+// ServiceAccountToken contains parameters of a token.
+type ServiceAccountToken struct {
+    Audiences []string
+    ExpirationSeconds *int64
+}
+```
+
+These three fields are all optional:
+
+- **`ServiceAccountToken.Audiences`**: will be set in `TokenRequestSpec`. All
+  the audiences will be validated to not be empty (i.e. kube apiserver). SP of
+  the CSI driver is supposed to send a `TokenReview` with at least one of the
+  audiences specified.
+
+- **`ServiceAccountToken.ExpirationSeconds`**: will be set in
+  `TokenRequestSpec`. The issued token may have a different duration, so the
+  `ExpirationTimestamp` in `TokenRequestStatus` will be passed to CSI driver.
+
+- **`RequiresRemount`**: should be only set when the mounted volumes by the
+  CSI driver have TTL and require re-validation on the token.
+
+  - **Note**: Remount means re-execution of `NodePublishVolume` in scope of
+    CSI and there is no intervening unmounts. If use this option,
+    `NodePublishVolume` should only change the contents rather than the
+    mount because container will not be restarted to reflect the mount
+    change. The period between remounts is 0.1s which is hardcoded as
+    `reconcilerLoopSleepPeriod` in volume manager. However, the rate
+    `TokenRequest` is not 10/s because it will be cached until expiration.
+
+The token will be bounded to the pod that the CSI driver is mounting volumes for
+and will be set in `VolumeContext`:
+
+```go
+"csi.storage.k8s.io/serviceAccount.token": token
+"csi.storage.k8s.io/serviceAccount.token.expiry": expirationTimestamp
```

Comment:
Is this the key format used by other attributes passed to CSI driver? It doesn't look like it matches the annotations API conventions:

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#label-selector-and-annotation-conventions

Reply:
yes, it doesn't follow the conventions for annotation but i would not consider it as annotation because it is just keys in the `VolumeContext` map. other attributes are using the same key format. e.g `csi.storage.k8s.io/pod.name`.